### PR TITLE
Girder & machine frame disassembly changes

### PR DESCRIPTION
**Pull request by kingofkosmos**
Old way:
Wrenching a girder (unsecured) -> secures the girder.
Wrenching a girder (secured) -> disassembles the girder.
Crowbarring a girder (secured) -> unsecures the girder.

**New way:**
Wrenching a girder (unsecured) -> secures the girder (same as before).
Wrenching a girder (secured) -> unsecures the girder.
Welding a girder (unsecured) -> disassembles the girder.

Also:
**Machine frames** can now be secured/unsecured with a wrench and disassembled with a welder. Old way was that they couldn't be unsecured at all, and disassembly was done with a wrench.

This way they're more consistent with nearly every other frame in the game, which have to be first unsecured with a wrench and then disassembled with a welder.

### CODE REVIEW

**Comment by kingofkosmos** [comment]
Also fixes an unreported issue about securing an unsecured girder. You could start securing it, push it away and it would secure itself after a while, while being a tile away from you.

**Comment by Miauw62** [comment]
Makes sense :+1:
Op 4-apr.-2015 20:15 schreef "kingofkosmos" <notifications@github.com>:

**Comment by chimersable** [comment]
Does it work correctly with the plasma cutter, thermite, or other wall mod effects?

File: code/game/objects/structures/girders.dm [hunk 1]
```diff
@@ -4,48 +4,55 @@
...
 					return
-				user << "<span class='notice'>You secured the girder!</span>"
-				var/obj/structure/girder/G = new (loc)
-				transfer_fingerprints_to(G)
+				user << "<span class='notice'>You disassemble the girder.</span>"
+				var/obj/item/stack/sheet/metal/M = new (loc, 2)
+				transfer_fingerprints_to(M)
 				qdel(src)
 
+	else if(istype(W, /obj/item/weapon/wrench) && state == 3 && 4)
```
Comment (phil235) [inline code/game/objects/structures/girders.dm]: why the `&& 4` ?

Comment (kingofkosmos) [inline code/game/objects/structures/girders.dm]: Good eye! I forgot it there after testing if I could remove 90% of the cultgirder's mostly-duplicatecode and change cultgirder into just a state, but soon realized that this stuff was too advanced for me. I'll remove it.

**Comment by kingofkosmos** [comment]
>Does it work correctly with the plasma cutter, thermite, or other wall mod effects?
Plasma cutter works but spawns only 1 metal per girder. I'll make it spawn 2 metal sheets, is that ok? I'm not sure if I can make it spawn a plasteel after slicing a reinforced girder.
Thermite doesn't work on girders. Jackhammering the girder works as normal.
I'm not sure what you mean by wall mod effects. Wall stuff is in a different file, unaffected by this.

**Comment by ACCount12** [comment]
Keep crowbar unsecuring please, because emergency toolbox has only crowbar and welder in it.

**Comment by MrPerson** [comment]
What does that have to do with anything? If you want the emergency toolbox to continue to be able to cut down walls, give it a welder

File: code/game/objects/structures/girders.dm [hunk 2]
```diff
@@ -4,48 +4,56 @@
...
-				user << "<span class='notice'>You dissasembled the girder!</span>"
-				new /obj/item/stack/sheet/metal(get_turf(src))
-				qdel(src)
-		else if(!anchored)
-			if (!istype(src.loc, /turf/simulated/floor))
-				usr << "<span class='danger'>A floor must be present to secure the girder!</span>"
-				return
-			playsound(src.loc, 'sound/items/Ratchet.ogg', 100, 1)
-			user << "<span class='notice'>Now securing the girder...</span>"
+	if(istype(W, /obj/item/weapon/weldingtool) && state == 3)
```
Comment (RemieRichards) [inline code/game/objects/structures/girders.dm]: replace these ```state == 3``` Checks with ```state == GIRDER_DISPLACED```

At the top of this file you'll want to add
```
#define GIRDER_NORMAL 0
#define GIRDER_REINF_STRUTS 1
#define GIRDER_REINF 2
#define GIRDER_DISPLACED 3
```
And then replace all the other checks with these too.

**Comment by ACCount12** [comment]
@MrPerson, it has a welder since  #8668.

**Comment by KorPhaeron** [comment]
So now it takes 3 steps/2 uses of welder fuel to take apart even basic walls?

I mean it adds consistency but it also adds extra steps to a simple action

**Comment by ACCount12** [comment]
@KorPhaeron, yes, that's why i dislike it too.

**Comment by optimumtact** [comment]
Who actually fully disassembles walls?

A little extra fuel usage isn't going to hurt in that case because it implies you're doing large scale construction, so you should've brought a welding tank with you

**Comment by PKPenguin321** [comment]
I dislike this. Deconstructing girders has been done with a wrench for ages, [literally even in this ancient fucking build,](http://i.imgur.com/1KjDHgC.jpg), and changing it would feel really inconsistent. Not to mention that now if you don't have welding protection and need to get through a quick wall you have to risk twice the eye damage for no real reason.
:+1: for the fix, big :-1: for the construction steps change.

**Comment by kingofkosmos** [comment]
I hope I did the stuff RR said right... Tested and everything worked correctly.

I really hope to hear what people think about false walls being non-instant to build (like every other thing in the game), mostly hoping that it's ok to add that to this PR and that it's not such a big balance change that it requires a whole new PR?

File: code/game/objects/structures/girders.dm [hunk 3]
```diff
@@ -1,51 +1,65 @@
...
-				return
-			playsound(src.loc, 'sound/items/Ratchet.ogg', 100, 1)
-			user << "<span class='notice'>Now securing the girder...</span>"
+	if(istype(W, /obj/item/weapon/weldingtool) && state == GIRDER_DISPLACED)
+		var/obj/item/weapon/weldingtool/WT = W
+		if(WT.remove_fuel(0,user))
+			playsound(src.loc, 'sound/items/Welder2.ogg', 50, 1)
+			user.visible_message("<span class='warning'>[user] disassembles the girder.</span>", \
+								"You start to disassemble the girder...", "You hear welding and clanking.")
+			var oldloc = src.loc
```
Comment (phil235) [inline code/game/objects/structures/girders.dm]: "var/oldloc" , not "var oldloc" please.

File: code/game/objects/structures/girders.dm [hunk 4]
```diff
@@ -1,51 +1,65 @@
...
+				M.add_fingerprint(user)
 				qdel(src)
 
+	else if(istype(W, /obj/item/weapon/wrench) && state == GIRDER_DISPLACED)
+		if (!istype(src.loc, /turf/simulated/floor))
+			usr << "<span class='warning'>A floor must be present to secure the girder!</span>"
+			return
+		playsound(src.loc, 'sound/items/Ratchet.ogg', 100, 1)
+		user << "<span class='notice'>You start securing the girder...</span>"
+		var oldloc = src.loc
```
Comment (phil235) [inline code/game/objects/structures/girders.dm]: see above comment

File: code/game/objects/structures/girders.dm [hunk 5]
```diff
@@ -95,23 +110,28 @@
 			if(/obj/item/stack/sheet/metal, /obj/item/stack/sheet/metal/cyborg)
 				if(!anchored)
 					if (S.use(2))
-						user << "<span class='notice'>You create a false wall! Push on it to open or close the passage.</span>"
+						user << "<span class='notice'>You start building a false wall...</span>"
+						var oldloc = src.loc
```
Comment (phil235) [inline code/game/objects/structures/girders.dm]: see above

File: code/game/objects/structures/girders.dm [hunk 6]
```diff
@@ -123,22 +143,27 @@
 			if(/obj/item/stack/sheet/plasteel)
 				if(!anchored)
 					if(S.amount < 2)
-						user << "<span class='warning'>You need at least two sheets to create a false wall.</span>"
+						user << "<span class='warning'>You need at least two sheets to create a false wall!</span>"
 						return
 					S.use(2)
-					user << "<span class='notice'>You create a false wall! Push on it to open or close the passage.</span>"
+					user << "<span class='notice'>You start building a reinforced false wall...</span>"
+					var oldloc = src.loc
```
Comment (phil235) [inline code/game/objects/structures/girders.dm]: see above

Comment (kingofkosmos) [inline code/game/objects/structures/girders.dm]: Check. Done to all of them.

**Comment by phil235** [comment]
> Not to mention that now if you don't have welding protection and need to get through a quick wall you have to risk twice the eye damage for no real reason.

Not really, you'd just use welder then wrench and move the displaced girder. No need to disassemble it if you only want to get through.

**Comment by KorPhaeron** [comment]
Change its layer var.

Also why any delay at all? What does a 2 second delay add besides slight annoyance and tedium? Doesn't add any interesting gameplay, doesn't address a balance issue, just a delay for the sake of one

**Comment by ACCount12** [comment]
>You're not supposed to be able to break down walls with the emergency toolbox, it's only supposed to help with airlock opening (crowbaring and unwelding) imo.

I added miniwelder specially for disassembling a wall or two in case of emegrency.

**Comment by kingofkosmos** [comment]
>Also why any delay at all? What does a 2 second delay add besides slight annoyance and tedium? Doesn't add any interesting gameplay, doesn't address a balance issue, just a delay for the sake of one

Mostly for the sake of consistency also since the majority of things take some time to build. Lightning-fast building feels out-of-place nowadays and makes no sense IC-wise either.

**Comment by norill** [comment]
machine frames use wrench to disassemble, but computer frames use welder.

**Comment by WJohn** [comment]
Machine frames should probably be anchorable with a wrench, and destroyed by a welder, then.

**Comment by kingofkosmos** [comment]
Already working on that one and it just needs testing. Should I add that to this PR or make a new one for it?

**Comment by phil235** [comment]
Feel free to add it to that PR.

**Comment by kingofkosmos** [comment]
Added. Didn't know how to make it so that machine frames are built unsecured and also when a SMES (or any machine) is taken apart, it would create a secured machine frame instead of an unsecured one. So they are now already secured when built.
Tried to implement the new if(do_after(user, 40, target = src))-thing, but didn't get it to work, so it's done the old fashioned way.

Stuff has been tested and seems to be working fine.

File: code/game/machinery/constructable_frame.dm [hunk 7]
```diff
@@ -59,27 +59,67 @@
...
 			if(istype(P, /obj/item/stack/cable_coil))
 				var/obj/item/stack/cable_coil/C = P
 				if(C.get_amount() >= 5)
 					playsound(src.loc, 'sound/items/Deconstruct.ogg', 50, 1)
-					user << "<span class='notice'>You start to add cables to the frame.</span>"
+					user << "<span class='notice'>You start to add cables to the frame...</span>"
+					var oldloc = src.loc
 					if(do_after(user, 20))
+						if(src.loc!=oldloc)
+							return
```
Comment (Jordie0608) [inline code/game/machinery/constructable_frame.dm]: do_after already handles this if you assign a target as one of the arguments.

File: code/game/machinery/constructable_frame.dm [hunk 8]
```diff
@@ -59,27 +59,67 @@
...
+				if(WT.remove_fuel(0,user))
+					playsound(src.loc, 'sound/items/Welder2.ogg', 50, 1)
+					user.visible_message("<span class='warning'>[user] disassembles the frame.</span>", \
+										"<span class='notice'>You start to disassemble the frame...</span>", "You hear welding and clanking.")
+					var oldloc = src.loc
+					if(do_after(user, 40))
+						if( !WT.isOn() )
+							return
+						if(src.loc!=oldloc)
+							return
```
Comment (Jordie0608) [inline code/game/machinery/constructable_frame.dm]: See above.

File: code/game/machinery/constructable_frame.dm [hunk 9]
```diff
@@ -59,27 +59,67 @@
...
 			if(istype(P, /obj/item/weapon/wrench))
+				user << "<span class='notice'>You start [anchored ? "un" : ""]securing [name]...</span>"
 				playsound(src.loc, 'sound/items/Ratchet.ogg', 75, 1)
-				user << "<span class='notice'>You dismantle the frame.</span>"
-				new /obj/item/stack/sheet/metal(src.loc, 5)
-				qdel(src)
+				var oldloc = src.loc
+				if(do_after(user, 40))
+					if(src.loc!=oldloc)
+						return
```
Comment (Jordie0608) [inline code/game/machinery/constructable_frame.dm]: See above.

File: code/game/machinery/constructable_frame.dm [hunk 10]
```diff
@@ -59,27 +59,67 @@
...
+					anchored = !anchored
+
 		if(2)
+			if(istype(P, /obj/item/weapon/wrench))
+				user << "<span class='notice'>You start [anchored ? "un" : ""]securing [name]...</span>"
+				playsound(src.loc, 'sound/items/Ratchet.ogg', 75, 1)
+				var oldloc = src.loc
+				if(do_after(user, 40))
+					if(src.loc!=oldloc)
+						return
```
Comment (Jordie0608) [inline code/game/machinery/constructable_frame.dm]: See above.

File: code/game/objects/structures/girders.dm [hunk 11]
```diff
@@ -1,51 +1,65 @@
...
+			var/oldloc = src.loc
 			if(do_after(user, 40))
-				if(!src.loc)
+				if( !WT.isOn() )
 					return
-				user << "<span class='notice'>You secured the girder!</span>"
-				var/obj/structure/girder/G = new (loc)
-				transfer_fingerprints_to(G)
+				if(src.loc!=oldloc)
+					return
```
Comment (Jordie0608) [inline code/game/objects/structures/girders.dm]: See above.

File: code/game/objects/structures/girders.dm [hunk 12]
```diff
@@ -1,51 +1,65 @@
...
+	else if(istype(W, /obj/item/weapon/wrench) && state == GIRDER_DISPLACED)
+		if (!istype(src.loc, /turf/simulated/floor))
+			usr << "<span class='warning'>A floor must be present to secure the girder!</span>"
+			return
+		playsound(src.loc, 'sound/items/Ratchet.ogg', 100, 1)
+		user << "<span class='notice'>You start securing the girder...</span>"
+		var/oldloc = src.loc
+		if(do_after(user, 40))
+			if(src.loc!=oldloc)
+				return
```
Comment (Jordie0608) [inline code/game/objects/structures/girders.dm]: See above.

File: code/game/objects/structures/girders.dm [hunk 13]
```diff
@@ -95,23 +110,28 @@
 			if(/obj/item/stack/sheet/metal, /obj/item/stack/sheet/metal/cyborg)
 				if(!anchored)
 					if (S.use(2))
-						user << "<span class='notice'>You create a false wall! Push on it to open or close the passage.</span>"
+						user << "<span class='notice'>You start building a false wall...</span>"
+						var/oldloc = src.loc
+						if(do_after(user, 20))
+							if(src.loc!=oldloc)
+								return
```
Comment (Jordie0608) [inline code/game/objects/structures/girders.dm]: See above.

File: code/game/objects/structures/girders.dm [hunk 14]
```diff
@@ -123,22 +143,27 @@
...
-						user << "<span class='warning'>You need at least two sheets to create a false wall.</span>"
+						user << "<span class='warning'>You need at least two sheets to create a false wall!</span>"
 						return
 					S.use(2)
-					user << "<span class='notice'>You create a false wall! Push on it to open or close the passage.</span>"
+					user << "<span class='notice'>You start building a reinforced false wall...</span>"
+					var/oldloc = src.loc
+					if(do_after(user, 20))
+						if(src.loc!=oldloc)
+							return
```
Comment (Jordie0608) [inline code/game/objects/structures/girders.dm]: See above.

File: code/game/objects/structures/girders.dm [hunk 15]
```diff
@@ -271,24 +298,29 @@
...
+	add_fingerprint(user)
+	if(istype(W, /obj/item/weapon/weldingtool))
+		var/obj/item/weapon/weldingtool/WT = W
+		if(WT.remove_fuel(0,user))
+			playsound(src.loc, 'sound/items/Welder2.ogg', 50, 1)
+			user.visible_message("<span class='warning'>[user] disassembles the girder.</span>", \
+								"<span class='notice'>You start to disassemble the girder...</span>", "You hear welding and clanking.")
+			if(do_after(user, 40))
+				if( !WT.isOn() )
+					return
```
Comment (Jordie0608) [inline code/game/objects/structures/girders.dm]: See above.

**Comment by Jordie0608** [comment]
Could you update the changelog for the machine frame changes.

**Comment by kingofkosmos** [comment]
All done!

File: code/game/machinery/constructable_frame.dm [hunk 16]
```diff
@@ -59,27 +59,55 @@
...
 					return
+			if(istype(P, /obj/item/weapon/weldingtool) && !anchored)
+				var/obj/item/weapon/weldingtool/WT = P
+				if(WT.remove_fuel(0,user))
+					playsound(src.loc, 'sound/items/Welder2.ogg', 50, 1)
+					user.visible_message("<span class='warning'>[user] disassembles the frame.</span>", \
+										"<span class='notice'>You start to disassemble the frame...</span>", "You hear welding and clanking.")
+					if(do_after(user, 40, target = src))
+						if( !WT.isOn() )
+							return
```
Comment (phil235) [inline code/game/machinery/constructable_frame.dm]: Need a check here to make sure state is still equal to 1 after the do_after ends.

File: code/game/machinery/constructable_frame.dm [hunk 17]
```diff
@@ -59,27 +59,55 @@
...
+						var/obj/item/stack/sheet/metal/M = new (loc, 5)
+						M.add_fingerprint(user)
+						qdel(src)
 			if(istype(P, /obj/item/weapon/wrench))
+				user << "<span class='notice'>You start [anchored ? "un" : ""]securing [name]...</span>"
 				playsound(src.loc, 'sound/items/Ratchet.ogg', 75, 1)
-				user << "<span class='notice'>You dismantle the frame.</span>"
-				new /obj/item/stack/sheet/metal(src.loc, 5)
-				qdel(src)
+				if(do_after(user, 40, target = src))
```
Comment (phil235) [inline code/game/machinery/constructable_frame.dm]: same here, we don't want unanchored machine_frame with circuitboard and wire.

File: code/game/machinery/constructable_frame.dm (hunk 16 above)
Comment (kingofkosmos) [inline code/game/machinery/constructable_frame.dm]: So just to make this clear, like this?
					if(do_after(user, 40, target = src))
						if( !WT.isOn() )
							return
						if(state == 1)
							user << "<span class='notice'>You disassemble the frame.</span>"
							var/obj/item/stack/sheet/metal/M = new (loc, 5)
							M.add_fingerprint(user)
							qdel(src)

Comment (phil235) [inline code/game/machinery/constructable_frame.dm]: yes

**Comment by Supermichael777** [comment]
feels weird that you need a welder to take apart something you didn't weld in the first place, but construction is weird. Also that picture of an old build MUH EYES

**Comment by kingofkosmos** [comment]
Checks added.

File: code/game/machinery/constructable_frame.dm (hunk 16 above)
Comment (phil235) [inline code/game/machinery/constructable_frame.dm]: You didn't add the check here.

Comment (kingofkosmos) [inline code/game/machinery/constructable_frame.dm]: Whoops, it's there now.

**Pull request merged**